imap/caldav_alarm.c

```diff
@@ -865,20 +838,169 @@ static void process_one_record(struct mailbox *mailbox, uint32_t imap_uid,
 
     syslog(LOG_DEBUG, "processing per-user alarms");
 
-    mailbox_get_annotate_state(mailbox, record.uid, NULL);
-    annotatemore_findall(mailbox->name, record.uid, PER_USER_CAL_DATA,
+    mailbox_get_annotate_state(mailbox, record->uid, NULL);
+    annotatemore_findall(mailbox->name, record->uid, PER_USER_CAL_DATA,
                          /* modseq */ 0, &process_peruser_alarms_cb,
                          &prock, /* flags */ 0);
 
     data.lastrun = runtime;
-    write_lastalarm(mailbox, &record, &data);
+    write_lastalarm(mailbox, record, &data);
 
-    update_alarmdb(mailbox->name, record.uid, data.nextcheck);
+    update_alarmdb(mailbox->name, record->uid, data.nextcheck);
 
 done_item:
     if (ical) icalcomponent_free(ical);
 }
 
+static void process_futurerelease(struct mailbox *mailbox,
+                                    struct index_record *record)
+{
+    struct buf buf = BUF_INITIALIZER;
+    json_t *submission = NULL, *identity, *envelope;
+    int r = 0;
+
+    syslog(LOG_DEBUG, "processing future delivery");
+
+    if ((record->internal_flags & FLAG_INTERNAL_EXPUNGED) ||
```

Comment:
This DELETED test could probably be lifted up into the upper loop - we're not going to want to process calendar events if the record has been marked DELETED either really, though it will probably never happen.